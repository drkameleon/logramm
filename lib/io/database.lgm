/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/io/database.lgm
 **********************************************************/

//================================================
// Imported modules
//================================================

import core;

//================================================
// Module definition
//================================================

Database :: 
{

	//================================================
	// External D-bindings
	//================================================

	external open(database) # [
		.params		: "string", 
		.returns    : "boolean", 
		.help		: "Opens/Creates database by name and returns 'true' if successful"
	] => "sqlitedb.open";

	external close(database) # [
		.params		: "string", 
		.returns    : "boolean", 
		.help		: "Closes database by name returns 'true' if successful"
	] => "sqlitedb.close";

	external query(database,sql) # [
		.params		: "string,string", 
		.returns    : "boolean", 
		.help		: "Executes SQL query to database and returns 'true' if successful"
	] => "sqlitedb.query";

	external nextRow(database) # [
		.params		: "string", 
		.returns    : "any", 
		.help		: "Returns dictionary of next row in result set, or 'false' if unsuccessful"
	] => "sqlitedb.nextRow";

	external allRows(database) # [
		.params		: "string", 
		.returns    : "array", 
		.help		: "Returns an array of rows with result set"
	] => "sqlitedb.allRows";

	external lastInsertId(database) # [
		.params		: "string", 
		.returns    : "number", 
		.help		: "Returns the id of the last inserted row"
	] => "sqlitedb.lastInsertId";

	//================================================
	// Functions
	//================================================

	create(database,table,keys,types) # [
		.params		: "string,string,array,array",
		.returns 	: "no",
		.help		: "Create table using 'keys' with 'types'"
	] =>
	{
		kvPairs = [];

		i = 0;
		@ i<count(keys) :
		{
			kvPairs += keys[i] + " " + types[i];
			i+=1;
		}

		kvPairsStr = join(kvPairs,", ");

		sql = "CREATE TABLE IF NOT EXISTS {:table} ({:kvPairsStr});";

		query(database,sql);
	}

	create(database,table,keytypes) # [
		.params		: "string,string,dictionary",
		.returns 	: "no",
		.help		: "Create table using 'keys' with 'types', with dictionary"
	] =>
	{
		keys = keys(keytypes);
		types = values(keytypes);
		
		create(database,table,keys,types);
	}

	insert(database,table,keys,values) # [
		.params		: "string,string,array,array",
		.returns 	: "number",
		.help		: "Inserts key/value pairs in table and returns id"
	] =>
	{
		newValues = [];

		@ val->values : newValues += "'" + val + "'";

		keysStr = join(keys,", ");
		valuesStr = join(newValues,", ");

		sql = "INSERT INTO {:table} ({:keysStr}) VALUES ({:valuesStr});";

		query(database,sql);

		==> lastInsertId(database);
	}

	insert(database,table,keyvalues) # [
		.params		: "string,string,dictionary",
		.returns 	: "number",
		.help		: "Inserts key/value pairs in table and returns id"
	] =>
	{
		keys = keys(keyvalues);
		values = values(keyvalues);

		==> insert(database,table,keys,values);
	}

	select(database,table,what,where) # [
		.params		: "string,string,array,array",
		.returns 	: "array",
		.help		: "Selects rows from table, based on conditions 'where', and returns result"
	] =>
	{
		 count(what)>1 : whatStr = join(what,", ");
				  else : whatStr = what[0];

		count(where)>0 : whereStr = " WHERE " + join(where," AND ");
				  else : whereStr = "";

		sql = "SELECT {:whatStr} FROM {:table} {:whereStr};";

		query(database,sql);

		==> allRows(database);
	}

	selectAll(database,table,where) # [
		.params		: "string,string,array",
		.returns 	: "array",
		.help		: "Selects all rows from table, based on conditions 'where', and returns result"
	] => select(database,table,["*"],where);

	delete(database,table,where) # [
		.params		: "string,string,array",
		.returns 	: "no",
		.help		: "Delete rows from table, based on conditions 'where'"
	] =>
	{
		count(where)>0 : whereStr = " WHERE " + join(where," AND ");
				  else : whereStr = "";

		sql = "DELETE FROM {:table} {:whereStr};";

		query(database,sql);
	}

	update(database,table,keys,values,where) # [
		.params		: "string,string,array,array,string",
		.returns 	: "no",
		.help		: "Updates rows in table, based on conditions 'where'"
	] =>
	{
		kvPairs = [];

		i = 0;
		@ i<count(keys) :
		{
			kvPairs += keys[i] + " = " + values[i];
			i+=1;
		}

		kvPairsStr = join(kvPairs,", ");

		count(where)>0 : whereStr = " WHERE " + join(where," AND ");
				  else : whereStr = "";

		sql = "UPDATE {:table} SET {:kvPairsStr} {:whereStr};";

		query(database,sql);
	}

	update(database,table,keyvalues,where) # [
		.params		: "string,string,dictionary,string",
		.returns 	: "no",
		.help		: "Updates rows in table, based on conditions 'where'"
	] =>
	{
		keys = keys(keyvalues);
		values = values(keyvalues);

		update(database,table,keyvalues,where);
	}
	
}

