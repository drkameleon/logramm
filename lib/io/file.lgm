/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/io/file.lgm
 **********************************************************/

//================================================
// Module definition
//================================================

File :: 
{

	//================================================
	// External D-bindings
	//================================================

	external read(filename) # [
		.params		: "string", 
		.returns    : "string", 
		.help		: "Returns contents of file as a string"
	] => "file.read";

	external write(filename,data) # [
		.params		: "string,string", 
		.returns    : "no", 
		.help		: "Writes data to file 'filename'"
	] => "file.write";

	external append(filename,data) # [
		.params		: "string,string", 
		.returns    : "no", 
		.help		: "Appends data to file 'filename'"
	] => "file.append";

	external rename(fileFrom,fileTo) # [
		.params		: "string,string", 
		.returns    : "no", 
		.help		: "Renames file to 'fileTo'"
	] => "file.rename";

	external delete(filename) # [
		.params		: "string", 
		.returns    : "no", 
		.help		: "Deletes file 'filename'"
	] => "file.deleteFile";

	external exists(filename) # [
		.params		: "string", 
		.returns    : "boolean", 
		.help		: "Check if file 'filename' exists"
	] => "file.exists";

}

