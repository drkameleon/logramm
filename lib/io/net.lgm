/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/io/net.lgm
 **********************************************************/

//================================================
// Module definition
//================================================

Net :: 
{
	//================================================
	// External D-bindings
	//================================================

	external get(url) # [
		.params		: "string", 
		.returns    : "string", 
		.help		: "Returns string contents of page at 'url'"
	] => "net.get";

	external post(url,data) # [
		.params		: "string,string", 
		.returns    : "string", 
		.help		: "Returns string contents of page at 'url', after an HTTP Post request with 'data'"
	] => "net.post";

	external download(url,filepath) # [
		.params		: "string,string", 
		.returns    : "no", 
		.help		: "Downloads file at 'url' to local 'filepath'"
	] => "net.download";

}

