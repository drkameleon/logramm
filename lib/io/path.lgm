/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/io/path.lgm
 **********************************************************/

//================================================
// Module definition
//================================================

Path :: 
{

	//================================================
	// External D-bindings
	//================================================

	external isFile(path) # [
		.params		: "string", 
		.returns    : "boolean", 
		.help		: "Checks if file at 'path' is a file"
	] => "path.isFile";

	external isDirectory(path) # [
		.params		: "string", 
		.returns    : "boolean", 
		.help		: "Checks if file at 'path' is a directory"
	] => "path.isDirectory";

	external isSymlink(path) # [
		.params		: "string", 
		.returns    : "boolean", 
		.help		: "Checks if file at 'path' is a symlink"
	] => "path.isSymlink";

	external createDirectory(path) # [
		.params		: "string", 
		.returns    : "no", 
		.help		: "Creates directory at 'path'"
	] => "path.createDirectory";

	external currentDirectory()	# [
		.params		: "", 
		.returns    : "string", 
		.help		: "Returns current working directory"
	] => "path.currentDirectory";

	external directoryContents(dir)	# [
		.params		: "string", 
		.returns    : "array", 
		.help		: "Returns array with contents of given directory"
	] => "path.directoryContents";

}
