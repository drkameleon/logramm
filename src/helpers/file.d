/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** src/helpers/file.d
 **********************************************************/

module helpers.file;

//================================================
// Imports
//================================================

import std.conv;
import std.random;
import std.stdio;
import std.string;

//================================================
// Functions
//================================================

string searchFileInPaths(string filename, string[] paths)
{
	foreach (string path; paths)
	{
		try 
		{
			
			string filepath = path ~ "/" ~ filename;

			auto file = File(filepath, "r");

			return filepath;
		}
		catch (Exception e)
		{
			
		}
	}

	return filename; // The file was NOT found. Return its name.
}

string tempWithContents(string contents)
{
	string filename = format("/tmp/tmp%s.lgm",to!string(uniform(0,10000)));
	auto f = File(filename, "w"); // open for writing
	f.write(contents);
	f.close();
	return filename;
}