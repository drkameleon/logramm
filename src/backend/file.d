/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** src/backend/file.d
 **********************************************************/

module backend.file;

//================================================
// Imports
//================================================

import std.array;
import std.conv;
import std.file;
import std.stdio;
import std.string;

import value;

//================================================
// Class
//================================================

class LGM_File
{
	static Value read(Value[] v)
	{
		string filename = v[0].content.s;

		string contents = to!string(std.file.read(filename));

		return new Value(contents);
	}

	static Value write(Value[] v)
	{
		string filename = v[0].content.s;
		string data = v[1].content.s;

		std.file.write(filename, data);

		return new Value(0);
	}

	static Value append(Value[] v)
	{
		string filename = v[0].content.s;
		string data = v[1].content.s;

		std.file.append(filename, data);

		return new Value(0);
	}

	static Value rename(Value[] v)
	{
		string filename = v[0].content.s;
		string new_filename = v[1].content.s;

		std.file.rename(filename, new_filename);

		return new Value(0);
	}

	static Value deleteFile(Value[] v)
	{
		string filename = v[0].content.s;

		std.file.remove(filename);

		return new Value(0);
	}

	static Value exists(Value[] v)
	{
		string filename = v[0].content.s;

		bool ret = std.file.exists(filename);

		return new Value(ret);
	}
}
