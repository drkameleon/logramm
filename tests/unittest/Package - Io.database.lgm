import io.database;
import io.file;

db = "unittest";
dbfile = db+".db";

// Creating the DB and some tables

out db.open();

keys = ["username","password"];
types = ["TEXT","TEXT"];

out db.create("users",keys,types);

keytypes = ["post":"TEXT","title":"TEXT"];

out db.create("posts",keytypes);

// Inserting elements

values = ["drkameleon","1234567"];

out db.insert("users",keys,values);
out db.insert("users",["username":"ozmax","password":"7654321"]);

keyvalues = ["post":"somePost","title":"someTitle"];

out db.insert("posts",keyvalues);

out db.selectAll("users", ["username='drkameleon'"]);
out db.selectAll("users", ["username='ozmax'"]);
out db.selectAll("posts", ["post='somePost'"]);
out db.selectAll("posts", ["post='anotherPost'"]);

out db.delete("posts",["post='somePost'"]);
out db.selectAll("users", []);
out db.selectAll("posts", []);

out dbfile.delete();
