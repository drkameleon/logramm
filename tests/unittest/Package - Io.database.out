true
0
0
1
2
1
[["username":"drkameleon", "password":"1234567"]]
[["username":"ozmax", "password":"7654321"]]
[["post":"somePost", "title":"someTitle"]]
[]
0
[["username":"drkameleon", "password":"1234567"],["username":"ozmax", "password":"7654321"]]
[]
0
